Hakeem Thomas <br>
Visualizing an EIS sweep from one row of data

---------------

## __Importing Libraries__

In [30]:
import pandas as pd              ##Open source library that focuses on data manipulation.
import plotly.express as px      ##Used for visualizing data.

### Creating the dataframe to manipulate

In [3]:
##Read the dataframe from excel
df = pd.read_excel('EIS_subset_data.xlsx')

In [4]:
##Showing the dataframe
df

Zre(ω1)  Zre(ω2)  Zre(ω3)  Zre(ω4)  Zre(ω5)  Zre(ω6)  Zre(ω7)  Zre(ω8)  \
0   0.3847  0.39156  0.39684  0.40341  0.40963  0.41925  0.42764   0.4353   

   Zre(ω9)  Zre(ω10)  ...  Zim(ω54)  Zim(ω55)  Zim(ω56)  Zim(ω57)  Zim(ω58)  \
0  0.44525   0.45744  ...   0.13411   0.15473   0.18024   0.21535   0.25411   

   Zim(ω59)  Zim(ω60)  Cycle Number  Capacity  Cell ID  
0   0.29026   0.32795             0  37.20271    25C01  

[1 rows x 123 columns]

Shows a dataframe with 123 columns, 120 pertaining to EIS data

In [12]:
##Seperate the EIS data only into a new dataframe
##Can use the drop function to remove the last three rows.
##Can be advantageous if the last three columns are going to be the same categories no matter what degree of sweep you use.

df1 = df.drop(['Capacity', 'Cell ID', 'Cycle Number'], axis = 1)  ##Only EIS data.

## 'axis = 1' signifies that the '.drop' function is looking at columns. 'axis = 0' would be for executing on rows.


In [31]:
#Show new dataframe
df1

Zre(ω1)  Zre(ω2)  Zre(ω3)  Zre(ω4)  Zre(ω5)  Zre(ω6)  Zre(ω7)  Zre(ω8)  \
0   0.3847  0.39156  0.39684  0.40341  0.40963  0.41925  0.42764   0.4353   

   Zre(ω9)  Zre(ω10)  ...  Zim(ω51)  Zim(ω52)  Zim(ω53)  Zim(ω54)  Zim(ω55)  \
0  0.44525   0.45744  ...   0.08888    0.1023   0.11767   0.13411   0.15473   

   Zim(ω56)  Zim(ω57)  Zim(ω58)  Zim(ω59)  Zim(ω60)  
0   0.18024   0.21535   0.25411   0.29026   0.32795  

[1 rows x 120 columns]

---------

### Pairing each real (Zre) and Imaginary (Zim) frequency (ω) together and plot.

In [23]:
a = df1.iloc[:, : 60].values  ##First 60 columns.
b = df1.iloc[:,-60:].values   ##Last 60 columns.

In [24]:
a

array([[0.3847 , 0.39156, 0.39684, 0.40341, 0.40963, 0.41925, 0.42764,
        0.4353 , 0.44525, 0.45744, 0.4667 , 0.4804 , 0.49228, 0.50688,
        0.52266, 0.53725, 0.55511, 0.57423, 0.59175, 0.61094, 0.6316 ,
        0.65332, 0.6764 , 0.69883, 0.72724, 0.75353, 0.7826 , 0.81236,
        0.84168, 0.87406, 0.90143, 0.92776, 0.94976, 0.9694 , 0.9839 ,
        0.99781, 1.00815, 1.01627, 1.02429, 1.02973, 1.03536, 1.04075,
        1.04617, 1.05071, 1.05612, 1.06154, 1.0674 , 1.07274, 1.07784,
        1.08463, 1.0928 , 1.1027 , 1.1119 , 1.12293, 1.1379 , 1.15879,
        1.17998, 1.19951, 1.22204, 1.25668]])

In [25]:
b

array([[-3.5130e-02, -1.7000e-02,  9.0000e-05,  1.4050e-02,  2.6120e-02,
         3.5030e-02,  4.4770e-02,  5.0550e-02,  5.9740e-02,  6.6080e-02,
         7.3370e-02,  8.1160e-02,  8.9550e-02,  9.4770e-02,  1.0115e-01,
         1.0537e-01,  1.1623e-01,  1.2077e-01,  1.2648e-01,  1.3371e-01,
         1.3902e-01,  1.4582e-01,  1.5104e-01,  1.5489e-01,  1.6106e-01,
         1.6641e-01,  1.6872e-01,  1.7167e-01,  1.6948e-01,  1.6454e-01,
         1.5485e-01,  1.4326e-01,  1.2989e-01,  1.1741e-01,  1.0340e-01,
         9.2360e-02,  8.2050e-02,  7.2660e-02,  6.6650e-02,  6.0920e-02,
         5.6500e-02,  5.4040e-02,  5.3710e-02,  5.3440e-02,  5.4500e-02,
         5.5720e-02,  5.9810e-02,  6.5720e-02,  7.1990e-02,  7.9030e-02,
         8.8880e-02,  1.0230e-01,  1.1767e-01,  1.3411e-01,  1.5473e-01,
         1.8024e-01,  2.1535e-01,  2.5411e-01,  2.9026e-01,  3.2795e-01]])

In [26]:
##Create a NEW dataframe and store the arrays into two columns
df_EIS = pd.DataFrame()

##Good practice to use new dataframes incase you want to use any specific one for future use. 

##Turn the nested array into a single array to be able to put into a column to be paired

Re_Z = a.flatten()             ##Makes the nested arrays into a single array to be put into a data frame.
Im_Z = b.flatten()

df_EIS['X Values'] = Re_Z
df_EIS['Y Values'] = Im_Z

In [27]:
df_EIS

X Values  Y Values
0    0.38470  -0.03513
1    0.39156  -0.01700
2    0.39684   0.00009
3    0.40341   0.01405
4    0.40963   0.02612
5    0.41925   0.03503
6    0.42764   0.04477
7    0.43530   0.05055
8    0.44525   0.05974
9    0.45744   0.06608
10   0.46670   0.07337
11   0.48040   0.08116
12   0.49228   0.08955
13   0.50688   0.09477
14   0.52266   0.10115
15   0.53725   0.10537
16   0.55511   0.11623
17   0.57423   0.12077
18   0.59175   0.12648
19   0.61094   0.13371
20   0.63160   0.13902
21   0.65332   0.14582
22   0.67640   0.15104
23   0.69883   0.15489
24   0.72724   0.16106
25   0.75353   0.16641
26   0.78260   0.16872
27   0.81236   0.17167
28   0.84168   0.16948
29   0.87406   0.16454
30   0.90143   0.15485
31   0.92776   0.14326
32   0.94976   0.12989
33   0.96940   0.11741
34   0.98390   0.10340
35   0.99781   0.09236
36   1.00815   0.08205
37   1.01627   0.07266
38   1.02429   0.06665
39   1.02973   0.06092
40   1.03536   0.05650
41   1.04075   0.05404
42   1.04617   0.05371
43   1.05071   0.05344
44   1.05612   0.05450
45   1.06154   0.05572
46   1.06740   0.05981
47   1.07274   0.06572
48   1.07784   0.07199
49   1.08463   0.07903
50   1.09280   0.08888
51   1.10270   0.10230
52   1.11190   0.11767
53   1.12293   0.13411
54   1.13790   0.15473
55   1.15879   0.18024
56   1.17998   0.21535
57   1.19951   0.25411
58   1.22204   0.29026
59   1.25668   0.32795

In [28]:
px.scatter(df_EIS, x = 'X Values', y = 'Y Values')  ##Produces a plot of the EIS